internal/commands/exec/command.go

```diff
@@ -165,6 +170,37 @@ func (c *Command) Run(args []string) (ret int) {
 		bearerToken = strings.TrimSpace(string(data))
 	}
 
+	addresses, grpcPort, err := c.parseDiscoveryAddresses()
+	if errors.Is(err, discoverySyntaxError) {
+		//old syntax
+		grpcPort = c.flagConsulXDSPort
```

Comment:
This should always be the case, the port that we pass at the end of `CONSUL_HTTP_ADDR` is always the `http` port, not grpc. The way the discover library works is that it hands back ip/dns addresses to use for client initialization, but you have to pass the http port individually.